### PR TITLE
insert block by fetcher

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: consensus/istanbul/core/core.go [hunk 1]
```diff
@@ -179,29 +179,62 @@ func (c *core) commit() {
...
 	// Try to get last proposal
-	if lastProposal == nil {
-		lastProposal, lastProposer = c.backend.LastProposal()
-		if lastProposal.Number().Cmp(newView.Sequence) > 0 {
-			newView = &istanbul.View{
-				Sequence: new(big.Int).Add(lastProposal.Number(), common.Big1),
-				Round:    new(big.Int),
+	lastProposal, lastProposer := c.backend.LastProposal()
+	if c.current == nil || lastProposal.Number().Cmp(c.current.Sequence()) >= 0 {
+		if c.current != nil {
```
Comment (alanchchen) [inline consensus/istanbul/core/core.go]: If we only do things when `c.current != nil`, why we check `c.current == nil` ? Just for printing logs ?

File: consensus/istanbul/core/core.go [hunk 2]
```diff
@@ -179,29 +179,62 @@ func (c *core) commit() {
...
-			logger.Trace("Catch up latest proposal", "number", lastProposal.Number().Uint64(), "hash", lastProposal.Hash())
 		}
+
+		logger.Trace("Catch up latest proposal", "number", lastProposal.Number().Uint64(), "hash", lastProposal.Hash())
+	} else if lastProposal.Number().Cmp(big.NewInt(c.current.Sequence().Int64()-1)) == 0 {
+		if round.Cmp(common.Big0) == 0 {
+			// same seq and round, don't need to start new round
+			return
+		} else if round.Cmp(c.current.Round()) < 0 {
+			logger.Warn("New sequence should not be smaller than current sequence", "new_seq", lastProposal.Number().Int64())
```
Comment (bailantaotao) [inline consensus/istanbul/core/core.go]: round? sequence?

File: consensus/istanbul/core/core.go [hunk 3]
```diff
@@ -179,29 +179,62 @@ func (c *core) commit() {
...
+	} else {
+		logger.Warn("New sequence should be larger than current sequence", "new_seq", lastProposal.Number().Int64())
+		return
+	}
+
+	var newView *istanbul.View
+	if roundChange {
+		newView = &istanbul.View{
+			Sequence: big.NewInt(c.current.Sequence().Int64()),
+			Round:    big.NewInt(round.Int64()),
```
Comment (alanchchen) [inline consensus/istanbul/core/core.go]: go-ethereum style:
`new(big.Int).Set(c.current.Sequeunce())`
`new(big.Int).Set(round)`
`big.NewInt(10000)` <--- only for constants

File: consensus/protocol.go [hunk 4]
```diff
@@ -48,8 +48,8 @@ type Protocol struct {
 
 // Broadcaster defines the interface to broadcast blocks and find peer
 type Broadcaster interface {
```
Comment (alanchchen) [inline consensus/protocol.go]: After this change. I think `Broadcaster` should have a new name.

File: consensus/istanbul/backend.go [hunk 5]
```diff
@@ -57,4 +58,13 @@ type Backend interface {
 
 	// LastProposal retrieves latest committed proposal and the address of proposer
 	LastProposal() (Proposal, common.Address)
+
+	// HasBlock checks if the combination of the given hash and height matches any existing blocks
+	HasBlock(hash common.Hash, number *big.Int) bool
```
Comment (alanchchen) [inline consensus/istanbul/backend.go]: We mixed `Proposal` with `Block` here. We should choose one.

File: consensus/istanbul/core/core.go [hunk 6]
```diff
@@ -179,29 +179,62 @@ func (c *core) commit() {
...
+		}
+	} else {
+		newView = &istanbul.View{
+			Sequence: new(big.Int).Add(lastProposal.Number(), common.Big1),
+			Round:    new(big.Int),
+		}
+		c.lastProposal = lastProposal
+		c.lastProposer = lastProposer
```
Comment (alanchchen) [inline consensus/istanbul/core/core.go]: Do we still need these two ?

File: consensus/istanbul/core/core.go (hunk 2 above)
Comment (markya0616) [reply consensus/istanbul/core/core.go]: round

File: consensus/istanbul/core/core.go (hunk 1 above)
Comment (markya0616) [reply consensus/istanbul/core/core.go]: OK

File: consensus/istanbul/backend/handler.go [hunk 7]
```diff
@@ -93,20 +92,12 @@ func (sb *backend) SetBroadcaster(broadcaster consensus.Broadcaster) {
 	sb.broadcaster = broadcaster
 }
 
-func (sb *backend) NewChainHead(block *types.Block) error {
+func (sb *backend) NewChainHead() error {
 	sb.coreMu.Lock()
```
Comment (bailantaotao) [inline consensus/istanbul/backend/handler.go]: Why not use RLock if no value is changed

File: consensus/istanbul/core/core.go (hunk 3 above)
Comment (markya0616) [reply consensus/istanbul/core/core.go]: OK

File: consensus/protocol.go (hunk 4 above)
Comment (markya0616) [reply consensus/protocol.go]: Will rename it in later PR

File: consensus/istanbul/backend.go (hunk 5 above)
Comment (markya0616) [reply consensus/istanbul/backend.go]: OK

File: consensus/istanbul/core/core.go (hunk 2 above)
Comment (markya0616) [reply consensus/istanbul/core/core.go]: round

Comment (bailantaotao) [reply consensus/istanbul/core/core.go]: Should the log amend to round? 
```logger.Warn("New sequence should not be smaller than current sequence",  ... ```

File: consensus/istanbul/core/core.go (hunk 6 above)
Comment (markya0616) [reply consensus/istanbul/core/core.go]: we can remove `lastProposal`

**Review by tailingchen — APPROVED** [review]

File: consensus/istanbul/backend/handler.go (hunk 7 above)
Comment (markya0616) [reply consensus/istanbul/backend/handler.go]: OK

**Review by bailantaotao — APPROVED** [review]
LGTM

**Review by alanchchen — APPROVED** [review]

File: consensus/istanbul/backend/backend.go [hunk 8]
```diff
@@ -239,6 +239,11 @@ func (sb *backend) CheckSignature(data []byte, address common.Address, sig []byt
 	return nil
 }
 
+// HasPropsal implements istanbul.Backend.HashBlock
```
Comment (yutelin) [inline consensus/istanbul/backend/backend.go]: wrong comment

**Pull request merged**